src/renderer/wg_engine/tvgWgGeometry.h

```diff
@@ -24,102 +24,434 @@
+            else if (vleng[mid] < len) left = mid + 1;
+            else right = mid - 1;
+        }
+        return right + 1;
+    }
+
+    // get min and max values of the buffer
+    inline void getMinMax(Point& pmin, Point& pmax) const {
+        if (vcount == 0) return;
+        pmax = pmin = vbuff[0];
+        for (size_t i = 1; i < vcount; i++) {
+            pmin = min(pmin, vbuff[i]);
+            pmax = max(pmax, vbuff[i]);
+        }
+    }
+
+    // update points distancess to the prev point and total length
+    inline void updateDistances() {
+        if (vcount == 0) return;
+        vdist[0] = 0.0f;
+        vleng[0] = 0.0f;
+        for (size_t i = 1; i < vcount; i++) {
+            vdist[i] = distance(vbuff[i-1], vbuff[i]);
+            vleng[i] = vleng[i-1] + vdist[i];
+        }
+    }
+
+    // close vertex buffer
+    inline void close() {
+        // check if last point is not to close to the first point
+        if (!tvg::zero(distance2(vbuff[0], last())))
+            append(vbuff[0]);
+        closed = true;
+    }
+
+    // append point
+    inline void append(const Point& p) {
+        vbuff[vcount] = p;
+        vcount++;
+    }
+
+    // append source vertex buffer in index range from start to end (end not included)
+    inline void appendRange(const WgVertexBuffer& buff, size_t start_index, size_t end_index) {
+        for (size_t i = start_index; i < end_index; i++)
+            append(buff.vbuff[i]);
+    }
+
+    // append circle (list of triangles)
+    inline void appendCircle(float radius) {
+        // get approx circle length
+        float clen = 2.0f * radius * MATH_PI;
+        size_t nsegs = std::max((uint32_t)(clen / 8), 16U);
+        // append circle
+        Point prev { std::sin(0.0f) * radius, std::cos(0.0f) * radius };
+        for (size_t i = 1; i <= nsegs; i++) {
+            float t = (2.0f * MATH_PI * i) / nsegs;
+            Point curr { std::sin(t) * radius, std::cos(t) * radius };
+            append(Point{0.0f, 0.0f});
+            append(prev);
+            append(curr);
+            prev = curr;
+        }
+    }
 
-    void appendPoint(WgPoint pt);
-    void appendCubic(WgPoint p1, WgPoint p2, WgPoint p3, size_t nsegs = 16);
+    // append cubic spline
+    inline void appendCubic(const Point& v0, const Point& v1, const Point& v2, const Point& v3) {
+        // get approx cubic length
+        float clen = distance(v0, v1) + distance(v1, v2) + distance(v2, v3);
+        size_t nsegs = (uint32_t)(clen / 8);
+        // append cubic
+        for (size_t i = 1; i <= nsegs; i++) {
+            // get cubic spline interpolation coefficients
+            float t = (float)i / nsegs;
+            float t0 = 1.0f * (1.0f - t) * (1.0f - t) * (1.0f - t);
+            float t1 = 3.0f * (1.0f - t) * (1.0f - t) * t;
+            float t2 = 3.0f * (1.0f - t) * t * t;
+            float t3 = 1.0f * t * t * t;
+            append(v0 * t0 + v1 * t1 + v2 * t2 + v3 * t3);
```

Comment:
maybe like in ```Point Bezier::at(float t) const``` ?
``
    auto ax = start * it + ctrl1 * t;
    auto bx = ctrl1 * it + ctrl2 * t;
    auto cx = ctrl2 * it + end * t;
    ax = ax * it + bx * t;
    bx = bx * it + cx * t;
    return ax * it + bx * t;
```